File: python/copass-management/src/copass_management/tools/update_trigger.py
from __future__ import annotations

from typing import Any, Dict, TYPE_CHECKING

if TYPE_CHECKING:
    from copass_management.registrar import ToolContext


async def update_trigger(ctx: "ToolContext", input: Dict[str, Any]) -> Any:
    trigger_id = str(input["trigger_id"])
    kwargs: Dict[str, Any] = {}
    if "event_type_filter" in input and input["event_type_filter"] is not None:
        kwargs["event_type_filter"] = str(input["event_type_filter"])
    if "rate_limit_per_hour" in input:
        v = input["rate_limit_per_hour"]
        if v is None:
            kwargs["clear_rate_limit"] = True
        else:
            kwargs["rate_limit_per_hour"] = int(v)
    if "filter_config" in input:
        v = input["filter_config"]
        if v is None:
            kwargs["clear_filter_config"] = True
        elif isinstance(v, dict):
            kwargs["filter_config"] = dict(v)
    trigger = await ctx.client.agents.triggers.update_by_id(
        ctx.sandbox_id, trigger_id, **kwargs,
    )
    return {"trigger": trigger}

File: typescript/packages/management/src/tools/update_trigger.ts
import type { UpdateTriggerRequest } from '@copass/core';

import type { ToolContext, ToolHandler } from '../registrar.js';

export const updateTrigger: ToolHandler = async (
  ctx: ToolContext,
  input: Record<string, unknown>,
) => {
  const triggerId = String(input.trigger_id);
  const patch: UpdateTriggerRequest = {};
  if (typeof input.event_type_filter === 'string') {
    patch.event_type_filter = input.event_type_filter;
  }
  if ('rate_limit_per_hour' in input) {
    const v = input.rate_limit_per_hour;
    if (v === null) {
      patch.clear_rate_limit = true;
    } else if (typeof v === 'number') {
      patch.rate_limit_per_hour = v;
    }
  }
  if ('filter_config' in input) {
    const v = input.filter_config;
    if (v === null) {
      patch.clear_filter_config = true;
    } else if (typeof v === 'object' && v !== null) {
      patch.filter_config = v as Record<string, unknown>;
    }
  }
  const trigger = await ctx.client.agents.triggers.updateById(
    ctx.sandboxId,
    triggerId,
    patch,
  );
  return { trigger };
};
